File: csharp/DesiArchitect.DesignPatterns/03_Observer/After/AnalyticsSubscriber.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Observer.After;

public class AnalyticsSubscriber : IOrderStatusSubscriber
{
    public void OnStatusChanged(Order order)
        => Console.WriteLine($"  Analytics tracked: Order {order.Id} -> {order.Status}");
}

File: typescript/src/03-observer/after/analytics-subscriber.ts
import type { IOrderStatusSubscriber } from './order-status-subscriber.js';
import type { Order } from '../../shared/models.js';

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

export class AnalyticsSubscriber implements IOrderStatusSubscriber {
  async onStatusChanged(order: Order): Promise<void> {
    await sleep(300);
    console.log(`  [After] Analytics tracked: Order ${order.id} -> ${order.status}`);
  }
}
